modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+      halt("DistributedBag Internal Error: 'elems' is nil");
+    }
+
+    if isFull {
+      halt("DistributedBag Internal Error: SegmentBlock is Full");
+    }
+
+    elems[size] = elt;
+    size = size + 1;
+  }
+
+  inline proc pop() : eltType {
+    if elems == nil {
+      halt("DistributedBag Internal Error: 'elems' is nil");
+    }
+
+    if isEmpty {
+      halt("DistributedBag Internal Error: SegmentBlock is Empty");
+    }
+
+    size = size - 1;
+    var elt = elems[size];
+    return elt;
+  }
+
+  proc BagSegmentBlock(type eltType, capacity) {
+    if capacity == 0 {
+      halt("DistributedBag Internal Error: Capacity is 0...");
+    }
+
+    cap = capacity;
+    elems = c_malloc(eltType, capacity);
+  }
+
+  proc BagSegmentBlock(type eltType, ptr, capacity) {
+    cap = capacity;
+    elems = ptr;
+    size = cap;
+  }
+
+  proc ~BagSegmentBlock() {
+    c_free(elems);
+  }
+}
+
+/*
+  A segment is, in and of itself an unrolled linked list. We maintain one per core
+  to ensure maximum parallelism.
+*/
+pragma "no doc"
+record BagSegment {
+  type eltType;
+
+  // Used as a test-and-test-and-set spinlock.
+  var status : atomic uint;
+
+  var headBlock : BagSegmentBlock(eltType);
+  var tailBlock : BagSegmentBlock(eltType);
+
+  var nElems : atomic uint;
+
+  inline proc isEmpty {
+    return nElems.read() == 0;
+  }
+
+  inline proc acquireWithStatus(newStatus) {
+    return status.compareExchangeStrong(STATUS_UNLOCKED, newStatus);
+  }
+
+  // Set status with a test-and-test-and-set loop...
+  inline proc acquire(newStatus) {
+    while true {
+      if currentStatus == STATUS_UNLOCKED && acquireWithStatus(newStatus) {
+        break;
+      }
+
+      chpl_task_yield();
+    }
+  }
+
+  // Set status with a test-and-test-and-set loop, but only while it is not empty...
+  inline proc acquireIfNonEmpty(newStatus) {
+    while !isEmpty {
+      if currentStatus == STATUS_UNLOCKED && acquireWithStatus(newStatus) {
+        if isEmpty {
+          releaseStatus();
+          return false;
+        } else {
+          return true;
+        }
+      }
+
+      chpl_task_yield();
+    }
+
+    return false;
+  }
+
+  inline proc isUnlocked {
+    return status.read() == STATUS_UNLOCKED;
+  }
+
+  inline proc currentStatus {
+    return status.read();
+  }
+
+  inline proc releaseStatus() {
+    status.write(STATUS_UNLOCKED);
+  }
+
+  inline proc transferElements(destPtr, n, locId = here.id) {
+    var destOffset = 0;
+    var srcOffset = 0;
+    while destOffset < n {
+      if headBlock == nil || isEmpty {
+        halt(here, ": DistributedBag Internal Error: Attempted transfer ", n, " elements to ", locId, " but failed... destOffset=", destOffset);
+      }
+
+      var len = headBlock.size;
+      var need = n - destOffset;
+      // If the amount in this block is greater than what is left to transfer, we
+      // cannot begin transferring at the beginning, so we set our offset from the end.
+      if len > need {

```

Comment:
Pretty sure this would be simpler using the array slice transfer, but this is OK if that doesn't work out for some reason.